codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketProtocolHandlerTest.java

```diff
@@ -85,68 +85,66 @@ public void testPingPongFlowControlWhenAutoReadIsDisabled() {
-        // It would be handle ONLY on the next 'channel.read()' call.
-        assertNull(channel.readInbound());
-        assertNull(channel.readOutbound());
-
-        // Cleanup
-        response1.release();
-        message2.release();
-        message3.release();
+        try (PongWebSocketFrame response1 = channel.readOutbound()) {
+            assertEquals(text1, response1.binaryData().toString(UTF_8));
+
+            // And - one requested message was handled and propagated inbound
+            try (TextWebSocketFrame message2 = channel.readInbound()) {
+                assertEquals(text2, message2.text());
+
+                // And - no more messages were handled or propagated
+                assertNull(channel.readInbound());
+                assertNull(channel.readOutbound());
+
+                // When
+                channel.read();
+
+                // Then - one requested message was handled and propagated inbound
+                try (TextWebSocketFrame message3 = channel.readInbound()) {
+                    assertEquals(text3, message3.text());
+
+                    // And - no more messages were handled or propagated
+                    // Precisely, ping frame 'text4' was NOT read or handled.
+                    // It would be handle ONLY on the next 'channel.read()' call.
+                    assertNull(channel.readInbound());
+                    assertNull(channel.readOutbound());
+                }
+            }
+        }
         assertFalse(channel.finish());
     }
 
     @Test
     public void testPongFrameDropFrameFalse() {
         EmbeddedChannel channel = new EmbeddedChannel(new WebSocketProtocolHandler(false) { });
 
-        PongWebSocketFrame pingResponse = new PongWebSocketFrame();
+        PongWebSocketFrame pingResponse = new PongWebSocketFrame(channel.bufferAllocator());
         assertTrue(channel.writeInbound(pingResponse));
 
         assertPropagatedInbound(pingResponse, channel);
 
-        pingResponse.release();
+        pingResponse.close();
         assertFalse(channel.finish());
     }
 
     @Test
     public void testPongFrameDropFrameTrue() {
         EmbeddedChannel channel = new EmbeddedChannel(new WebSocketProtocolHandler(true) { });
 
-        PongWebSocketFrame pingResponse = new PongWebSocketFrame();
+        PongWebSocketFrame pingResponse = new PongWebSocketFrame(channel.bufferAllocator());
         assertFalse(channel.writeInbound(pingResponse)); // message was not propagated inbound
     }
 
     @Test
     public void testTextFrame() {
         EmbeddedChannel channel = new EmbeddedChannel(new WebSocketProtocolHandler() { });
 
-        TextWebSocketFrame textFrame = new TextWebSocketFrame();
+        TextWebSocketFrame textFrame = new TextWebSocketFrame(channel.bufferAllocator());
         assertTrue(channel.writeInbound(textFrame));
 
         assertPropagatedInbound(textFrame, channel);
 
-        textFrame.release();
+        textFrame.close();
```

Comment:
We wrote this message to a channel, but we're still allowed to close it out here?

Reply:
Apparently EmbeddedChannel don't close messages like that.